### ISSUE TITLE
Memory issue

### ISSUE THREAD

**Issue by SerjPridestar** [MetalPetal/MetalPetal#49]
Hello Yuao! Hope you are doing good!

As you might remember, I was experimenting using MetalPetal to generate a video asset from a set of images. I added a feature to it where on completion the class returns all the images it used to render the video for any post processing video needs. I used `let cgImages = try images.map({ try self.context.makeCGImage(from: $0) })` after that code is executed I notice a huge memory footprint remaining in memory. I'm not sure if I'm doing anything wrong, but I checked the leak profiler and there was no leak detected.

Usually when I am generating a video out of images the render method increments a 15mb - 20mb per image. After the video generation is complete I am left with a 600mb memory in RAM, before I would be left with approximately 140mb, which is also bad, however, it was easier to get by with. Additionally, it's important for me to mention that I am rendering really high resolution images directly from the iOS camera, they are approximately 3k or 4k resolution. You might be wondering why not just transform the images to a smaller resolution and render them then? That's because it's a requirement for me to be able to render really high resolution images as I would want the video to be high resolution itself.

I also tried checking the allocation profiler, but there is an unusual case where the allocation shows the memory spike to 30mb when the video generation begins and then it gradually goes down to 10mb; on the other hand, on the leak profiler actually shows the spike to 400mb 600mb. It is very weird.

I have set up the project [here](https://github.com/SerjPridestar/MetalPetalVideoGenerator) if you are willing to do some investigations. I have set warnings where I think things are going wrong so that it would be easy to navigate to the problem directly. Thank you in advance for reading this long issue.

**Comment by YuAo** [comment]
Try add `context.reclaimResources()` after `makeCGImage`?

**Comment by SerjPridestar** [comment]
I tried adding it still the problem persists after I finish generating all the CGImages:

```
let cgImages = try images.map({ try self.context.makeCGImage(from: $0) })
let uiImages = cgImages.map({ UIImage(cgImage: $0) })
self.context.reclaimResources()
```

I don't think I have different sizes in my render pipeline, since I take all the images and crop and transform them all to the same size and then call the `context.render(image, to: pixelBuffer)`. I did this for the same exact reason you mentioned, I wanted the rendering to be more performant and use up less memory and reuse the resources already allocated. After a lot of experimentation I noticed by having all the images the same size this is the case. However, still after each `context.render` call I see resources up to 15mb - 20mb. I will try adding `context.reclaimResources()` after each render call see if I still see this progressive incremental increase in memory.

In general my goal out of using `Metal` with the `AVAssetWriterInput` is to be able to generate really high quality videos with no constraints on how many images I could render on the `AVAssetWriterInputPixelBufferAdaptor`.

**Comment by YuAo** [comment]
The situation here is a little bit weird, I may need some more time to investigate.

**Comment by SerjPridestar** [comment]
Alright! If you need anything or I could do anything to help please do inform me! Again thank you very much!

**Comment by SerjPridestar** [comment]
@YuAo I'll give it a shot right now. Good job digging and finding the problem, I spent a day on it and couldn't find the main problem. 

Is it the same with the `context.render(image, to: pixelBuffer)`? The call to that increments a gradual increase in memory. Maybe I'm wrong, but to me there shouldn't be a limitation on how many images you can draw to the `AVAssetWriterInputPixelBufferAdaptor`.  I'm guessing the process is so fast that the system isn't able to flush any memory needed, thus an overload of images just crashes the app before even getting the chance to clear up memory. I would love to hear your thoughts on this. Thank you very much Yuao your work is very much appreciated!

**Comment by YuAo** [comment]
Yes, it is the same with `context.render(image, to: pixelBuffer)`.

**Comment by SerjPridestar** [comment]
@YuAo I also did some extra digging, also having the `PHCachingImageManager.requestImage` makes a significant difference in memory. It seems that UIImage is an uncompressed version of the Data, so dealing with 4k resolution images this can make a major difference. Now I'm using `PHCachingImageManager.requestImageData` instead. The downside of this, is you use up more processing time orienting the images to the correct orientation. Its a good trade off in my opinion, much less memory used to a one second more processing.

Moreover, I guess it should be the same when having the exporter complete I should return a compressed version of the images instead of a `UIImage`. 

By the way Yuao I was thinking of adding `MetalPetalVideoGenerator` to the MetalPetal repository as example on how to use MetalPetal to make videos out of images. What do you think?

**Comment by SerjPridestar** [comment]
Hello Yuao! I tried the improvements and great news the memory drops significantly and everything works as expected! Before the improvements after finishing exporting the memory would stay at around 324 mb, after improvements the memory is reclaimed and the memory is at 115 mb.

I think it is safe to close this issue for now!

Moreover, I agree about your opinion about adding `MetalPetalVideoGenerator` to the demos as its main focus is `AVFoundation` as its end goal is to make a video out of different sources of images. 

Once again great work with this and thank you for patching a fix so promptly much appreciated!

**Comment by SerjPridestar** [comment]
Only thing I noticed right now as you suggested instead of returning `CGImage/UIImage`s, I tried `CIImage`, not sure if it is a compressed version, but memory stays at 800 mb; on the other hand, `CGImage` after export has its memory stay at 400 mb. I did make sure to reclaim resources after 1.1s after generating the `CIImage`s

**Comment by YuAo** [comment]
`CIImage` may contains compressed or uncompressed image data.
The suggestion is to use compressed image data, like JPEG.

```Swift
let imageData: [Data] = try images.map({ item in
                                try autoreleasepool(invoking: {
                                    var img: CGImage?
                                    let task = try self.context.startTask(toCreate: &img, from: item, sRGB: false)
                                    task.waitUntilCompleted()
                                    let data = UIImage(cgImage: img!).jpegData(compressionQuality: 0.8)
                                    self.context.reclaimResources()
                                    return data!
                                })
                            })
```

**Comment by SerjPridestar** [comment]
@YuAo do we still need to call `self.context.reclaimResources()` after 1.1 seconds?

**Comment by SerjPridestar** [comment]
@YuAo for some reason if I don't reclaim resources after 1.1 seconds after generating the compressed images the memory isn't dropping. However, for the `try self.context.render(image, to: pixelBuffer)` it seems to be working fine.

**Comment by SerjPridestar** [comment]
It's all good now. After a lot of experimentation, I realized keeping a reference to the `MTIImage` is what was making the memory increment and not being able to `reclaimResources()`. So when I got done rendering the image to the video and tried to create the compressed image data, there wasn't enough memory to do that. Thus, I realized its best after I'm done with rendering the `MTIImage` I directly create the image data save it in an array and set the reference to that `MTIImage` to nil and reclaim the resources. Now it works perfectly! Thank you so much Yuao!!

**Comment by YuAo** [comment]
A `MTIImage`'s `cachePolicy` tells the `MTIContext` how it should handle the rendered buffer of that image. You can keep the `MTIImage` around, just make sure its cache policy is `.transient`.

**Comment by SerjPridestar** [comment]
@YuAo  Eventhough I had all the `MTIImage`s using the cache policy `.transient` having them in memory around they were using up a large amount of memory.

I changed the whole algorithm to be on demand basis. So I only create an `MTIImage` only when the time comes to render that `ImageData`.

In my opinion I don't I think it's `MTIImage` using up the memory per-say rather than the `CIImage`.  What I noticed while using the `MTIImage(ciImage: ciImage, isOpaque: true)` constructor is that `MTIImage` rather the `MTICIImagePromise` keeps a reference to the `CIImage` as you mentioned above `CIImage` may contain compressed or uncompressed image data. In my use case I have 4k images using that constructor the `CIImage` does definitely occupy a large space in memory and is costly. At the end of the day, acting like resources aren't scare while programming is just lazy programming.
